### DISCUSSION TITLE
Understand recursion with examples

### DISCUSSION THREAD

**Discussion by Charles-depot** [pamelafox/recursive-visualizations#21] [Show and tell]
Hi Pamela,
Thanks for recursive-visualisations.
I am a mathematics teacher and computer teacher in France in a high school.
I use recursive-visualizations to make students understand the notion of recursion.
Below is the 3-part course and additional examples.
The course is not yet complete and uses Jupyter notebooks

you can find examples using recursion visualizer here --> https://nuage03.apps.education.fr/index.php/s/5WW4pgzSWJGTybR

You can use [Basthon](https://basthon.fr/) to view notebooks
## Récursivité (1/3) : notions de fonctions
This first activity aims to review and complete some notions on function calls and to get started with some simple graphical visualization tools: Python Tutor and Recursion Visualizer.
Below are the examples using recursion visualizer in notebook [Récursivité (1/3)](https://nuage03.apps.education.fr/index.php/s/5WW4pgzSWJGTybR)

[example 1](https://www.recursionvisualizer.com/?function_definition=def%20f%28x%29%3A%0A%20%20%20%20return%202*x%20%2B%201%0A%0Adef%20ajouter%28x%29%3A%0A%20%20%20%20return%202%20%2B%20f%28x%29&function_call=ajouter%284%29)
[example 2](https://www.recursionvisualizer.com/?function_definition=def%20f%28x%29%3A%0A%20%20%20%20return%202*x%20%2B%201%0A%0Adef%20dilatation%28x%29%3A%0A%20%20%20%20return%20f%283*x%29&function_call=dilatation%284%29)
[example 3](https://www.recursionvisualizer.com/?function_definition=def%20f%28x%29%3A%0A%20%20%20%20return%202*x%20%2B%201%0A%0Adef%20g%28x%29%3A%0A%20%20%20%20return%20x**2%0A%0Adef%20somme%28x%29%3A%0A%20%20%20%20return%20f%28x%29%20%2B%20g%28x%29&function_call=somme%284%29)
[example 4](https://www.recursionvisualizer.com/?function_definition=def%20f%28x%29%3A%0A%20%20%20%20return%202*x%20%2B%201%0A%0Adef%20composition%28x%29%3A%0A%20%20%20%20return%20f%28f%28x%29%29&function_call=composition%284%29)
[example 5](https://www.recursionvisualizer.com/?function_definition=def%20f%28x%29%3A%0A%20%20%20%20return%202*x%20%2B%201%0A%0Adef%20calcul_terme_avec_f%28n%29%3A%0A%20%20%20%20u%20%3D%201%0A%20%20%20%20for%20i%20in%20range%281%2C%20n%2B1%29%3A%0A%20%20%20%20%20%20%20%20u%20%3D%20f%28u%29%0A%20%20%20%20return%20u&function_call=calcul_terme_avec_f%284%29)
[example 6](https://www.recursionvisualizer.com/?function_definition=def%20f%28x%29%3A%0A%20%20%20%20return%202*x%20%2B%201%0A%0Adef%20calcul_myst%C3%A9rieux_avec_f%28n%29%3A%0A%20%20%20%20if%20n%20%3D%3D%200%3A%0A%20%20%20%20%20%20%20%20return%201%0A%20%20%20%20else%3A%0A%20%20%20%20%20%20%20%20return%20f%28calcul_myst%C3%A9rieux_avec_f%28n-1%29%29&function_call=calcul_myst%C3%A9rieux_avec_f%284%29)

## Récursivité (2/3) : notions de récursivité
The purpose of this second activity is to handle some classic examples related to recursion and to identify some important notions.
Below are the examples using recursion visualizer in notebook [Récursivité 2/3](https://nuage03.apps.education.fr/index.php/s/5WW4pgzSWJGTybR)

[example 1](https://www.recursionvisualizer.com/?function_definition=def%20fact%28n%29%3A%0A%20%20%20%20assert%20isinstance%28n%2C%20int%29%20and%20n%3E%3D0%0A%20%20%20%20if%20n%20%3D%3D%200%3A%0A%20%20%20%20%20%20%20%20return%201%0A%20%20%20%20else%3A%0A%20%20%20%20%20%20%20%20return%20n*fact%28n-1%29&function_call=fact%284%29)
[example 2](https://www.recursionvisualizer.com/?function_definition=def%20myst%C3%A8re%28n%29%3A%0A%20%20%20if%20n%20%3E%205%3A%0A%20%20%20%20%20%20%20return%201%0A%20%20%20else%3A%0A%20%20%20%20%20%20%20return%20n%20%2B%20myst%C3%A8re%28n%2B1%29&function_call=myst%C3%A8re%282%29)
[example 3](https://www.recursionvisualizer.com/?function_definition=def%20myst%C3%A8re_infini%28n%29%3A%0A%20%20%20if%20n%20%3C%3D%202%3A%0A%20%20%20%20%20%20%20return%201%0A%20%20%20else%3A%0A%20%20%20%20%20%20%20return%20n%20%2B%20myst%C3%A8re_infini%28n%2B1%29&function_call=myst%C3%A8re_infini%283%29)
[example 4](https://www.recursionvisualizer.com/?function_definition=def%20carre%28n%29%3A%0A%20%20%20%20assert%20isinstance%28n%2C%20int%29%20and%20n%20%3E%3D%200%0A%20%20%20%20if%20n%20%3D%3D%200%3A%0A%20%20%20%20%20%20%20%20return%200%0A%20%20%20%20return%20carre%28n-1%29%20%2B%202*n%20-%201%0A%0A&function_call=carre%284%29)
[example 5](https://www.recursionvisualizer.com/?function_definition=def%20maximum%28liste%29%3A%0A%20%20%20%20if%20len%28liste%29%20%3D%3D%201%3A%0A%20%20%20%20%20%20%20%20return%20liste%5B0%5D%0A%20%20%20%20indice_milieu%20%3D%20len%28liste%29%2F%2F2%0A%20%20%20%20max_gauche%20%3D%20maximum%28liste%5B%3Aindice_milieu%5D%29%0A%20%20%20%20max_droite%20%3D%20maximum%28liste%5Bindice_milieu%3A%5D%29%0A%20%20%20%20if%20max_gauche%20%3E%20max_droite%3A%0A%20%20%20%20%20%20%20%20return%20max_gauche%0A%20%20%20%20return%20max_droite%0A%0A%0A%0A&function_call=maximum%28%5B2%2C5%2C6%2C4%2C9%2C8%2C1%5D%29
)

## NSI Récursivité (3/3) : mise en pratique

This last activity aims to put into practice what has been seen on recursion.
Below are the examples using recursion visualizer in notebook [Récursivité 3/3](https://nuage03.apps.education.fr/index.php/s/5WW4pgzSWJGTybR)
[example 1](https://www.recursionvisualizer.com/?function_definition=def%20puissance_rec%28x%2Cn%29%3A%0A%20%20%20%20if%20n%20%3D%3D%200%3A%0A%20%20%20%20%20%20%20%20return%201%0A%20%20%20%20if%20n%20%3D%3D%201%3A%0A%20%20%20%20%20%20%20%20return%20x%0A%20%20%20%20else%3A%0A%20%20%20%20%20%20%20%20return%20x*puissance_rec%28x%2Cn-1%29&function_call=puissance_rec%282%2C5%29)
[example 2](https://www.recursionvisualizer.com/?function_definition=def%20FiboRec%28n%29%3A%0A%20%20%20%20if%20n%20%3C%3D%201%20%3A%0A%20%20%20%20%20%20%20%20return%20n%0A%20%20%20%20else%3A%0A%20%20%20%20%20%20%20%20return%20FiboRec%28n-1%29%20%2B%20FiboRec%28n-2%29&function_call=FiboRec%285%29)

## Other examples

[example 1](https://www.recursionvisualizer.com/?function_definition=def%20valMaxTab%28T%20%2Ci%20%2C%20j%20%29%3A%0A%20%20%20%20if%20i%20%3D%3D%20j%3A%0A%20%20%20%20%20%20%20%20return%20T%5Bi%5D%0A%20%20%20%20else%3A%0A%20%20%20%20%20%20%20%20m%20%3D%20%28i%2Bj%29%2F%2F2%0A%20%20%20%20%20%20%20%20a%20%3D%20valMaxTab%28T%2C%20i%2C%20m%29%0A%20%20%20%20%20%20%20%20b%20%3D%20valMaxTab%28T%2C%20m%2B1%2C%20j%29%0A%20%20%20%20%20%20%20%20return%20max%28a%2Cb%29%0A%0A%0A&function_call=valMaxTab%28%5B7%2C5%2C4%2C3%2C2%2C12%5D%2C0%2Clen%28%5B7%2C5%2C4%2C3%2C2%2C12%5D%29-1%29)
[example 2](https://www.recursionvisualizer.com/?function_definition=def%20pair%28n%29%20%3A%0A%20%20%20%20if%20n%20%3D%3D%200%3A%0A%20%20%20%20%20%20%20%20return%20True%0A%20%20%20%20else%3A%0A%20%20%20%20%20%20%20%20return%20impair%28n-1%29%0A%0Adef%20impair%28n%29%3A%0A%20%20%20%20if%20n%20%3D%3D%200%3A%0A%20%20%20%20%20%20%20%20return%20False%0A%20%20%20%20else%3A%0A%20%20%20%20%20%20%20%20return%20pair%28n-1%29%0A&function_call=pair%285%29)
[example 3](https://www.recursionvisualizer.com/?function_definition=def%20arbre_additif%28valeurs%29%3A%0A%20%20%20%20if%20len%28valeurs%29%20%3D%3D%201%3A%0A%20%20%20%20%20%20%20%20return%20%28None%2C%20valeurs%5B0%5D%2C%20None%29%0A%20%20%20%20%0A%20%20%20%20m%20%3D%20len%28valeurs%29%20%2F%2F%202%0A%20%20%20%20a%20%3D%20arbre_additif%28valeurs%5B%3Am%5D%29%0A%20%20%20%20b%20%3D%20arbre_additif%28valeurs%5Bm%3A%5D%29%0A%20%20%20%20return%20%28a%2C%20a%5B1%5D%2Bb%5B1%5D%2C%20b%29&function_call=%20arbre_additif%28%5B3%2C%205%2C%207%2C%204%5D%29)
[example 4](https://www.recursionvisualizer.com/?function_definition=def%20fib_mem%28n%29%3A%0A%20%20mem%20%3D%20%5B0%5D*%28n%2B1%29%20%20%23permet%20de%20cr%C3%A9er%20un%20tableau%20contenant%20n%2B1%20z%C3%A9ro%0A%20%20return%20fib_mem_c%28n%2Cmem%29%0A%0Adef%20fib_mem_c%28n%2Cm%29%3A%0A%20%20if%20n%3D%3D0%20or%20n%3D%3D1%3A%0A%20%20%20%20m%5Bn%5D%3Dn%0A%20%20%20%20return%20n%0A%20%20elif%20m%5Bn%5D%3E0%3A%0A%20%20%20%20return%20m%5Bn%5D%0A%20%20else%3A%0A%20%20%20%20m%5Bn%5D%3Dfib_mem_c%28n-1%2Cm%29%20%2B%20fib_mem_c%28n-2%2Cm%29%0A%20%20%20%20return%20m%5Bn%5D&function_call=fib_mem%285%29)
[example 5](https://www.recursionvisualizer.com/?function_definition=def%20maximum%28liste%29%3A%0A%20%20%20%20if%20len%28liste%29%3D%3D1%3A%0A%20%20%20%20%20%20%20%20return%20liste%5B0%5D%0A%20%20%20%20indice_milieu%20%3D%20len%28liste%29%2F%2F2%0A%20%20%20%20max_g%20%3D%20maximum%28liste%5B%3Aindice_milieu%5D%29%0A%20%20%20%20max_d%20%3D%20maximum%28liste%5Bindice_milieu%3A%5D%29%0A%20%20%20%20if%20max_g%20%3E%20max_d%3A%0A%20%20%20%20%20%20%20%20return%20max_g%0A%20%20%20%20return%20max_d%0A%0A%0A%0A&function_call=maximum%28%5B2%2C5%2C6%2C4%2C9%2C8%2C4%2C2%2C10%2C5%2C9%5D%29)

**Comment by pamelafox** [comment]
Thanks so much for sharing! Great examples here of building up intuition for call graphs with non-recursive examples first.
